File: packages/provide-uterm-server/src/provide/uterm/server/webhook_signing.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Shared signing helpers for governance webhook payloads."""

from __future__ import annotations

import hashlib
import hmac
import time
from typing import cast

_DEFAULT_MAX_AGE_S = 300.0


def build_webhook_signature(secret: str, body: bytes, timestamp: str) -> str:
    """Return ``sha256=<hex>`` over ``"{timestamp}.".encode() + body`` (replay-resistant)."""
    signed = timestamp.encode("utf-8") + b"." + body
    digest = hmac.new(secret.encode("utf-8"), signed, hashlib.sha256).hexdigest()
    return f"sha256={digest}"


def verify_webhook_signature(
    secret: str | None,
    body: bytes,
    signature_header: str | None,
    timestamp_header: str | None,
    *,
    max_age_s: float = _DEFAULT_MAX_AGE_S,
    now: float | None = None,
) -> bool:
    """Verify ``X-Uterm-Signature`` over ``ts.body`` and that the timestamp is fresh."""
    # L8a: fail closed when there is no signing key. A signature cannot be
    # authenticated without a shared secret — verifying with an empty key would
    # HMAC over an empty key, which any attacker who knows the body+timestamp
    # can forge. Reject before touching the signature so an empty-key HMAC can
    # never validate, regardless of how this function is reached (a directly
    # constructed WebhookIdentityProvider bypasses the config validator).
    if not (secret or "").strip():
        return False
    if not signature_header or not timestamp_header:
        return False
    try:
        ts_val = float(timestamp_header)
    except (TypeError, ValueError):
        return False
    current = time.time() if now is None else now
    if abs(current - ts_val) > max_age_s:
        return False
    supplied = signature_header.strip()
    if supplied.lower().startswith("sha256="):
        supplied = supplied.split("=", 1)[1].strip()
    if not supplied:
        return False
    # The fail-closed guard above guarantees secret is a non-empty str here.
    expected = build_webhook_signature(cast("str", secret), body, timestamp_header).split("=", 1)[1]
    return hmac.compare_digest(supplied, expected)

File: packages/provide-uterm-ts/src/serverauth/webhook-signing.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Signing helpers for governance webhook payloads.
 *
 * Port of the Python module `provide.uterm.server.webhook_signing`.
 *
 * Signed over `"{timestamp}.{body}"`, so a captured request cannot be
 * replayed once the window has passed — and verified fail-closed when there
 * is no signing secret, because an empty key HMACs to something anyone who
 * knows the body and the timestamp can forge.
 */

import { createHmac } from "node:crypto";
import { digestsMatch } from "./digests.ts";

/** How far a timestamp may be from now, in either direction. */
export const WEBHOOK_MAX_AGE_S = 300.0;

/** The `sha256=<hex>` signature over `timestamp` and `body`. */
export function buildWebhookSignature(secret: string, body: Uint8Array, timestamp: string): string {
  // The separator matters: concatenating the two would let a body that
  // starts with digits masquerade as a different timestamp.
  const signed = Buffer.concat([Buffer.from(`${timestamp}.`, "utf8"), Buffer.from(body)]);
  return `sha256=${createHmac("sha256", Buffer.from(secret, "utf8")).update(signed).digest("hex")}`;
}

/**
 * Whether a webhook request carries a fresh, valid signature.
 *
 * @returns `false` for every failure. There is no partial success here: a
 *   request either proved it came from the holder of the secret or it did not.
 */
export function verifyWebhookSignature(
  secret: string | undefined,
  body: Uint8Array,
  signatureHeader: string | undefined,
  timestampHeader: string | undefined,
  options: { maxAgeS?: number; now?: number } = {},
): boolean {
  // Before anything else: a signature cannot be authenticated without a
  // shared secret, and HMAC with an empty key is forgeable by anyone. This
  // has to hold however the function is reached, including from a directly
  // constructed provider that never saw the config validator.
  if ((secret ?? "").trim() === "") {
    return false;
  }
  if (
    signatureHeader === undefined ||
    signatureHeader === "" ||
    timestampHeader === undefined ||
    timestampHeader === ""
  ) {
    return false;
  }
  // Trimmed first: an empty or whitespace header parses as zero, which would
  // be a timestamp from 1970 and so outside every window — but relying on
  // that is relying on an accident.
  const trimmedTimestamp = timestampHeader.trim();
  const timestamp = Number(trimmedTimestamp);
  if (trimmedTimestamp === "" || !Number.isFinite(timestamp)) {
    return false;
  }
  const now = options.now ?? Date.now() / 1000;
  const maxAge = options.maxAgeS ?? WEBHOOK_MAX_AGE_S;
  // Both directions: a clock ahead of the sender is as much of a replay
  // window as one behind.
  if (Math.abs(now - timestamp) > maxAge) {
    return false;
  }
  let supplied = signatureHeader.trim();
  if (supplied.toLowerCase().startsWith("sha256=")) {
    supplied = supplied.slice(supplied.indexOf("=") + 1).trim();
  }
  // Not load-bearing on its own — an empty string is a different length
  // from a digest, so the comparison refuses it either way — but a bare
  // prefix is a malformed header, not a wrong signature, and saying so here
  // does not depend on that.
  if (supplied === "") {
    return false;
  }
  const expected = buildWebhookSignature(secret as string, body, timestampHeader).split("=")[1] as string;
  return digestsMatch(supplied, expected);
}
